### DISCUSSION TITLE
Distribute zipped Windows binaries?

### DISCUSSION THREAD

**Discussion by rgwood** [dankamongmen/notcurses#2151] [Ideas]
I'd love to start messing around with Notcurses on Windows. But I've found MSYS2 to be a barrier - I had some install issues and tbh it just feels a bit alien to me as a Windows developer. I totally get the appeal for *nix-centric projects and developers, that's just not me.

However, I don't have any real need to build from source - I'd be fine with prebuilt DLLs+EXEs. The package manager situation on Windows is... complicated, but one simple option might be to just publish a zip of the Windows build output with each GitHub release.

P.S. Notcurses is incredibly cool and I love the demos you've built for it. Keep up the fantastic work :)

**Comment by dankamongmen** [comment]
aye, i intend to distribute Windows binaries as soon as i understand the Windows ecosystem and experience better (here, as i have many times, i will point out the irony of my employment at Microsoft). windows dev has been very strange in my recent experience. like, there's no shared lib folder that everyone develops against? you just copy your own libs into your own directory? and then assume everything you build against is in that build directory? then redistribute the libs you need? madness.

i would recommend you to #2071, but you just commented there =]. #1966 is probably also relevant.

thank you for the kind words, and i'll be frank: windows knowledge is in very, very short supply on this project. #1966 wants a Visual Studio project file and some directions. put that together for us, and you've just become an essential contributor. and with that, i'd have a much better understanding of what and how to install.

like, if people are building against their own copies, i'd assume you just want DLLs in a ZIP, which you can unpack where you'd like. do you even typically have an installer for libraries in this environment?

help me rgwood i have no idea what i'm doing i am eight

**Reply by dankamongmen** [reply]
but yes, the upshot is that i'd like to have windows binaries built and uploaded as artifacts on each tag. i'm just unsure what form to provide them as. ought i statically link them, so the MSYS2 components are provided directly? etc.

**Reply by rgwood** [reply]
It is madness, you're not wrong :). I don't have as many answers for you as I'd like, my own Windows experience is heavy on the .NET side of things and not so much on the C/C++ side (which would be much more useful for Notcurses). 

> ought i statically link them, so the MSYS2 components are provided directly? 

That's a good question - I would prefer static linking for simplicity but maybe other people have good reasons for needing dynamic linking?

I'll try to help out with Windows issues to the best of my ability, still need to do a lot of learning about the project though. Probably going to start out building Notcurses in WSL+Windows Terminal, then look at native Windows builds next.

**Reply by dankamongmen** [reply]
building in WSL ought be no problem whatsoever. you're in an entirely unix environment there.

inside, you ought use `TERM=ms-terminal`.

**Reply by WSLUser** [reply]
> Inside, you ought use TERM=ms-terminal.

I can't keep stressing, that the WT devs want you to use `xterm-256color `as that terminfo entry mentioned is ridiculously outdated. 

Secondly,
There are some older Win32 programs which do make use of shared libs that are installed in a general location, typically under a system protected directory, which applications use of via env vars. So you can host your dlls somewhere like Program Files (x86) (assuming this is a 64 bit library of course - which this should be ) and set the env var to your dlls so that any application that wants to make use of this library can reference the env var or they can specify the path directly, whichever works faster for them. 

Modern applications typically bundle all dependencies themselves **_even when there are other applications using the same dependencies_** due to issues that have arisen due to Windowsisms. They could make it work but no effort has really been seen there, at least from a user perspective.

**Reply by dankamongmen** [reply]
are they going to get the `ms-terminal` entry updated? they don't implement the entirety of `xterm-256color`  =[

**Reply by WSLUser** [reply]
Also, I would really really love to see a binary produced that has nothing to do with Msys2. Frankly, it's a blocker for most Windows users as there will an impression of needing to install Msys2, even though we both know isn't really needed, we just need the binaries produced from an msys2 build. More importantly though, I still believe a native toolchain can yield better compilation results than the current method but as you stated in my issue, we need to link the appropriate headers and libraries to ensure functionality isn't dropped. There are plenty of projects which have done full ports of their Linux application to Windows and maintain both. But since these projects all vary in nature, determining from those what this library in particular needs is the challenge. I said take a look at PDCurses before as it's the closest example to notcurses in providing a similar functionality that does work on all 3 major platforms. I would like to believe everything they're using for Windows would be all you need but this focuses on graphics a fair bit more than they do so maybe there's more dependencies to handle as result, especially if we want to tie into DirectX for rendering the graphics produced by this library through a TUI app (this last part might have everything to do with the terminal itself and nothing to do with this library).

**Reply by dankamongmen** [reply]
> Also, I would really really love to see a binary produced that has nothing to do with Msys2. Frankly, it's a blocker for most Windows users as there will an impression of needing to install Msys2, even though we both know isn't really needed, we just need the binaries produced from an msys2 build. More importantly though, I still believe a native toolchain can yield better compilation results than the current method but as you stated in my issue, we need to link the appropriate headers and libraries to ensure functionality isn't dropped. There are plenty of projects which have done full ports of their Linux application to Windows and maintain both. But since these projects all vary in nature, determining from those what this library in particular needs is the challenge. I said take a look at PDCurses before as it's the closest example to notcurses in providing a similar functionality that does work on all 3 major platforms. I would like to believe everything they're using for Windows would be all you need but this focuses on graphics a fair bit more than they do so maybe there's more dependencies to handle as result, especially if we want to tie into DirectX for rendering the graphics produced by this library through a TUI app (this last part might have everything to do with the terminal itself and nothing to do with this library).

i mean, i would like it too, but what alternative do you see to rewriting a ton of shit? and given that i'm not any authority on windows, probably rewriting it badly?

**Reply by dankamongmen** [reply]
one thing we *could* do is disrtribute the MSYS2 runtime with the source, and then a good visual studio project file ought be able to handle it transparently, i would think? or it could even fetch them, again transparently.

the user and probably the hacker needn't know they exist. but i'm not getting into the business of recreating them.

**Reply by dankamongmen** [reply]
i could even make a submodule repo that holds windows DLLs on which we depend, and the windows build could populate it. regular unix people could just ignore it.

**Reply by dankamongmen** [reply]
but if you want working POSIX code reimplemented wholesale with Windows code....i mean, just to keep my own limited sanity, and not have two distinct paths everywhere i touch an os resource, i'd end up reinventing cygwin *anyway*.

**Reply by dankamongmen** [reply]
i like supporting windows, and am eager to do so as best i can, but have no misconceptions -- if this project was a liferaft floating around in the northern atlantic wherein we dined on a dwindling supply of hi-c ecto cooler, and there weren't enough ecto coolers and assfloat to go around, i'm kicking windows out the boat. it's not my fault they looked at the great vista (pun not intended) of POSIX interfaces and said "nahhhh son we need capital letters DWORD is bond".

they. *we*, i mean. damnit, nicholas, *we*.

**Reply by rgwood** [reply]
Yup, that all makes sense. Windows is the odd one out here and I'm fully behind any approach that lets you support Windows while still maintaining your sanity. Personally I think an MSYS2-first build is reasonable even if it's not the most "native" way of doing things.

I might try to build Notcurses with MSBuild and Visual Studio sometime , but I am not optimistic about my chances of success (partially due to missing knowledge on my part, partially because working with MSBuild setup tends to be... draining).

**Reply by WSLUser** [reply]
> are they going to get the `ms-terminal` entry updated? they don't implement the entirety of `xterm-256color` =[

Well there's an issue for them to update it. I know they plan to at some point but they likely want several architectural design changes first such as their text buffer 2021 issue or creating ConPTY pass-through mode to enable VT sequences such as DCS to work for terminals that take a dependency on ConPTY.

**Reply by WSLUser** [reply]
Also if they're missing a feature that hasn't already been filed, they welcome issues tied to use-cases. If notcurses needs something Windows doesn't yet have,  let them know and someone will get to it eventually and of course they'll accept contributions as they do intend on being completely compatible with that term within reason.
